Source/OxyPlot/Series/ContourSeries.cs

```diff
@@ -486,100 +459,118 @@ private void AddContourLabels(Contour contour, ScreenPoint[] pts, ICollection<Co
-        /// <returns>The connected segment, or <c>null</c> if no segment was found.</returns>
-        private ContourSegment FindConnectedSegment(DataPoint point, double contourLevel, double eps, out bool reverse)
-        {
-            reverse = false;
-            foreach (var s in this.segments)
-            {
-                if (!AreClose(s.ContourLevel, contourLevel, eps))
-                {
-                    continue;
-                }
-
-                if (AreClose(point, s.StartPoint, eps))
-                {
-                    return s;
-                }
-
-                if (AreClose(point, s.EndPoint, eps))
-                {
-                    reverse = true;
-                    return s;
-                }
-            }
-
-            return null;
-        }
-
         /// <summary>
         /// Joins the contour segments.
         /// </summary>
         /// <param name="eps">The tolerance for segment ends to connect (squared distance).</param>
         private void JoinContourSegments(double eps = 1e-10)
         {
-            // This is a simple, slow, naïve method - should be improved:
-            // http://stackoverflow.com/questions/1436091/joining-unordered-line-segments
             this.contours = new List<Contour>();
-            var contourPoints = new List<DataPoint>();
-            int contourPointsCount = 0;
 
-            ContourSegment firstSegment = null;
-            int segmentCount = this.segments.Count;
-            while (segmentCount > 0)
+            static IEnumerable<SegmentPoint> GetPoints(ContourSegment segment)
             {
-                ContourSegment segment1 = null, segment2 = null;
+                var p1 = new SegmentPoint(segment.StartPoint);
+                var p2 = new SegmentPoint(segment.EndPoint);
+                p1.Partner = p2;
+                p2.Partner = p1;
+                yield return p1;
+                yield return p2;
+            }
 
-                if (firstSegment != null)
+            foreach (var group in this.segments.GroupBy(p => p.ContourLevel))
+            {
+                var level = group.Key;
+                var points = group.SelectMany(GetPoints).OrderBy(p => p.Point.X).ToList();
+
+                // first, go through the sorted points, find identical points and join them together 
+                for (var i = 0; i < points.Count - 1; i++)
                 {
-                    bool reverse;
+                    var currentPoint = points[i];
+                    if (currentPoint.Join != null)
+                    {
+                        continue;
+                    }
 
-                    // Find a segment that is connected to the head of the contour
-                    segment1 = this.FindConnectedSegment(contourPoints[0], firstSegment.ContourLevel, eps, out reverse);
-                    if (segment1 != null)
+                    var maxX = currentPoint.Point.X + eps;
+                    var i2 = 1;
```

Comment:
I'd prefer that `i2 = i + 1` rather than being an offset (simplifies everything a little).

Reply:
Unless the values are expected to be identical, sorting by coordinates will never be perfect. I'd use a quad-tree for this, though that might be overkill.

Reply:
Yes, sorting won't be perfect, but good enough for 'non-pathological' use cases - under normal circumstances sorting by X should pair corresponding points almost completely.